File: python/day15.py
a = 699
b = 124
factors = {'a': 16807, 'b': 48271}
divisor = 2147483647


def generator(value, factor, multi=1):
    while True:
        value = value * factor % divisor
        if value % multi == 0:
            yield value & 0xFFFF


part1_a = generator(a, factors['a'])
part1_b = generator(b, factors['b'])

part2_a = generator(a, factors['a'], 4)
part2_b = generator(b, factors['b'], 8)

first = sum(next(part1_a) == next(part1_b) for _ in range(40000000))
second = sum(next(part2_a) == next(part2_b) for _ in range(5000000))

print(first)
print(second)

File: ocaml/day15.ml
let rec generate v factor m =
  let divisor = 2147483647 in
  let nv = ref (v * factor) in
  while !nv >= divisor do
    nv := (!nv land divisor) + (!nv lsr 31)
  done;
  if !nv land (m - 1) = 0 then !nv
  else generate !nv factor m

let solve iter mA mB =
  let fA = 16807 in
  let fB = 48271 in
  let a = ref 699 in
  let b = ref 124 in
  let tot = ref 0 in
  for _ = 1 to iter do
    a := generate !a fA mA;
    b := generate !b fB mB;
    if !a land 0xffff = !b land 0xffff then incr tot
  done;
  !tot |> Printf.printf "%d\n"


let first = solve 40_000_000 1 1
let second = solve 5_000_000 4 8
